### PR TITLE
wxWidgets-3.2: update version to 3.1.3

### PR DESCRIPTION
**Pull request by ra1nb0w**
#### Description

bump wxWidgets-3.2 version to 3.1.3

require bump revision of the following port:
- freedv-gui (updated)
- limesuite
- wxMaxima (there is a new version - livecheck doesn't work)
- mediainfo
- saga (there is a new version - livecheck doesn't work)
- CubicSDR (updated with upstream fix)

others?

###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
macOS 10.15 19A602
Xcode 11.1 11A1027

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [ ] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by mojca** [comment]
Btw, there's also #4158.

**Comment by mojca** [comment]
Generally looks OK.

It would also be nice to increase the version number for `wxWidgets-common`, and maybe use the opportunity to check if there's anything special needed (or might be simplified) for < 10.9 now that MacPorts 2.6 compiler handling came out.

I would also like to test the latest version of `audacity` (and `audacity-devel`) which we should attempt to link against wxWidgets-3.2 if that's not the case yet.

If it's ok to have a grace period until November 6th to check a few more things  ...

**Comment by ra1nb0w** [comment]
> It would also be nice to increase the version number for `wxWidgets-common`,

done plus note on wxWidgets-3.2

> and maybe use the opportunity to check if there's anything special needed (or might be simplified) for < 10.9 now that MacPorts 2.6 compiler handling came out.

I don't have macOS < 10.9

> I would also like to test the latest version of `audacity` (and `audacity-devel`) which we should attempt to link against wxWidgets-3.2 if that's not the case yet.

audacity has a new version 2.3.2 so probably we can upgrade and link to wxwidgets-3.2 as default.
@RJVB  what do you think about removing -devel?

**Comment by mojca** [comment]
We can keep the devel for audacity and let it point to the latest commit, hoping that we could get rid of as many patches as possible (we have waaaaay too many). I plan to look into it this week starting from the latest changes from @RJVB.

**Comment by ra1nb0w** [comment]
Ok

**Comment by RJVB** [comment]
On Tuesday October 29 2019 02:44:15 rainbow wrote:

>> I would also like to test the latest version of `audacity` (and `audacity-devel`) which we should attempt to link against wxWidgets-3.2 if that's not the case yet.
>
>audacity has a new version 2.3.2 so probably we can upgrade and link to wxwidgets-3.2 as default.
>@RJVB  what do you think about removing -devel?

My own version of port:audacity is already at 2.3.2 (and 2.3.2.316 for the -devel port, which I use, or 2.3.2.535 on Linux). I link/run against a late wxW. 3.1.2 version (3.1.2.916) on both platforms.
Part of the reason for having the -devel port is to make patch maintenance easier, and that includes not having to add patches if upstream fixes something between releases.

I'll try to find some time to upgrade my own audacity-devel port for Mac, at least bringing it to the same version as the Linux version. No promise (rather a prediction that I will likely not get around to this).

R.

**Comment by cjones051073** [comment]
> Btw, there's also #4158.

I think we should look to merge these two PRs together, and get the both merged, once and for all.

**Comment by mojca** [comment]
> > Btw, there's also #4158.
> 
> I think we should look to merging these two PRs together, and get the both merged, once and for all.

Except that the other one may need changes now that we have a C++11-capable compiler almost everywhere anyway.

**Comment by cjones051073** [comment]
> > > Btw, there's also #4158.
> > 
> > 
> > I think we should look to merging these two PRs together, and get the both merged, once and for all.
> 
> Except that the other one may need changes now that we have a C++11-capable compiler almost everywhere anyway.

That is why I said merge the two PRs. 

Mostly likely the minimum that is needed is to enable c++11 with ` --enable-cxx11` and then instead of using the cxx11 PG using the new base support `compiler.cxx_standard     2011` in wxWidgets, and perhaps also in any ports using it.

**Comment by ra1nb0w** [comment]
enabled c++11

**Comment by cjones051073** [comment]
Thanks. Have you checked though to see if dependents now also need to specify that they require c++11 ? If wxWidgets exposes this, then this should be done in all dependents.

**Comment by ra1nb0w** [comment]
what do you mean with "checked"?

**Comment by cjones051073** [comment]
I guess what I mean is does enabling c++11 here force the use of this standard in dependents. If it does, every dependent should also specify this standard as well. I don’t know the port in question here well enough to know if this is the case or not.

**Comment by ra1nb0w** [comment]
but, what do I need to check? and How?
The issue I know is during linking the symbol table between different standard/compiler and generally the error arise at build stage. I tested mediainfo without problem.

**Comment by RJVB** [comment]
>but, what do I need to check? and How?
>The issue I know is during linking the symbol table between different standard/compiler and generally the error arise at build stage. I tested mediainfo without problem.

@Mojca, any idea if wxWidgets exposes the C++ standard to clients? If `mediainfo` builds and doesn't use C++11 (or later) itself then that should be sufficient proof that there is no need to activate C++ in *every* dependent.
But that's something that could be handled in the PG, anyway.

**Comment by michaelld** [comment]
This PR supersedes #4158, so I'm going to close out the other one in favo[u]r of this one.

**Comment by michaelld** [comment]
I'll echo @RJVB 's comment, which I believe was my experience from working on #4158 : this wx can itself use C++11, but it doesn't technically require C++11 to build; nor does it pass on the requirement for C++11 via its API to ports that depend on it.

**Comment by michaelld** [comment]
@ra1nb0w With this update in place, if any of the ports that depend on it that -do not- require C++11 for building successfully build and execute, then the API doesn't pass through the C++11 requirement. I strongly believe this is the case, and since there aren't that many dependent ports (like 4-6), it's pretty simple to test / verify.

**Comment by michaelld** [comment]
I updated CubicSDR (and LimeSuite-devel) ... so the former will just be a rev-bump now & the latter has to be fixed (sorry!).

**Comment by mojca** [comment]
My past experience from 3.0 has been that you could not really build C++11 software against wxWidgets which wasn't compiled with C++11 support. And vice versa. Maybe things have changed  in version 3.1, but I'm a bit skeptic. That said, one would not notice the issue on 10.9 and newer since C++11 is "kind of" enabled by default, and now we also support C++11 on 10.6 and newer.

This still leaves 10.5 though, but I'm not sure to what extent we can reasonably support those. I believe that 3.1 requires 10.8 SDK (and can thus be built on 10.7 provided the SKD is there), while 3.0 still works on 10.5.

**Comment by michaelld** [comment]
@kencu figured out how to get Wx 3.1 to build on 10.6 and 10.7: https://github.com/macports/macports-ports/pull/4158#issuecomment-486389092 ...

**Comment by michaelld** [comment]
I think Wx 3.1 doesn't work with 10.5 or 10.4 at all, so we really don't have to worry about their C++11 support or lack thereof, at least with respect to this PR.

**Comment by RJVB** [comment]
On Friday November 01 2019 07:41:49 Mojca Miklavec wrote:
>My past experience from 3.0 has been that you could not really build C++11 software against wxWidgets which wasn't compiled with C++11 support. And vice versa. 

With all due respect, I don't think that makes sense as long as a C++IJ-enabled build doesn't expose C++IJ features via public API, and given that there is only 1 C++ runtime library. Oh, and as long as symbol name mangling doesn't change as a function of C++ standard (but I don't think that's the case either).

Then again I'm not that much of an expert in C++...

**Comment by mojca** [comment]
What happened, at least with wxWidgets 3.0, was that configure script would check which headers were available. It tested whether `<foo>` was there; if not, it would check for `<tr1/foo>`, it would find that header in the pre-C++11 compiler, and then write `#define HAVE_TR1_FOO 1` to the config header that's always loaded whenever some piece of software uses wxWidgets. Then you tried to build something in C++11 mode, wxWidgets still insisted in loading `<tr1/foo>`, but that header was no longer discoverable when you asked the compiler to build something as C++11.

Anyway, it's pointless to discuss this ad infinity. Let's just always explicitly build wxWidgets-3.2 with C++11 enabled and fix other dependencies later as problems arise.

I'm thinking of unconditionally using wxWidgets 3.2 for Audacity for now. I have rebased their patches for wxWidgets 3.1.3, but there's a really massive chunk of code devoted to accessibility, which would make sense to upstream to wxWidgets. I'm pondering whether or not to include those patches here.

**Comment by ra1nb0w** [comment]
@michaelld revision bump done; what is the problem with limesuite-devel? on my macOS 10.15.1 build fine.

@mojca: how do you want to proceed with audacity?

**Comment by michaelld** [comment]
> @michaelld revision bump done; what is the problem with limesuite-devel? on my macOS 10.15.1 build fine.

No issues now. Before you were doing a full update to limesuite-devel ... in order to get your commit there to fix the Wx build. Now that that commit is already in the port, you've moved to a rev-bump. This all works for me & seems like a reasonable way to go.

**Comment by ra1nb0w** [comment]
ok, yes it is cleaner and the way to go.

**Comment by mojca** [comment]
I have an impression that `include/wx/osx/private/datatransfer.h` is being referenced, but not installed. This is what I get when installing the development version of Audacity:
```
/opt/local/Library/Frameworks/wxWidgets.framework/Versions/wxWidgets/3.1/include/wx-3.1/wx/osx/private.h:5:10: fatal error: 'wx/osx/private/datatransfer.h' file not found
#include "wx/osx/private/datatransfer.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by mojca** [comment]
Looks like the problem I'm experiencing has been introduced a month ago by https://github.com/wxWidgets/wxWidgets/pull/1264

**Comment by mojca** [comment]
wxMaxima's dependency recode seems to be failing with
```
checking whether yytext is a pointer... yes
checking for a Python interpreter with version >= 3.5... none
configure: error: no suitable Python interpreter found
Command failed:  cd "/opt/local/var/macports/build/_Users_runner_runners_2.159.2_work_1_s_textproc_recode/recode/work/recode-3.7.6" && ./configure --prefix=/opt/local 
Exit code: 1
```

**Comment by ra1nb0w** [comment]
> wxMaxima's dependency recode seems to be failing with

seems that recode needs python3/cython3 now
https://github.com/rrthomas/recode/commit/89237f2eeb57332d104148f25e119aea2d8308e0

**Comment by ra1nb0w** [comment]
> Looks like the problem I'm experiencing has been introduced a month ago by [wxWidgets/wxWidgets#1264](https://github.com/wxWidgets/wxWidgets/pull/1264)

mmm, `private` ;-)
anyway, we wait upstream response.

**Comment by RJVB** [comment]
BTW, any interest in changing to the CMake build system as part of this effort? It seems to offer all the configuration options the autoconf-based system offers, and should result in a faster build (esp. when using ninja) and benefit the port maintainers (easy out-of-tree building, importing the project directory into an IDE with full/better support for code parsing, ...)

**Comment by mojca** [comment]
When talking about CMake: are we talking about wxWidgets or Audacity?
(Audacity is definitely not ready with CMake just yet.)

I was actually wrong: it seems that I can neither build audacity nor audacity-devel with wxWidgets 3.1.3.

**Comment by RJVB** [comment]
>When talking about CMake: are we talking about wxWidgets or Audacity?

wxWidgets.

I've been tinkering with that on Linux: it works, but
- installs headers for all flavours (android, mswin, etc)
- installs non-versioned headers. I guess that might in fact a plus if you're installing into versioned subprefices...

>I was actually wrong: it seems that I can neither build audacity nor audacity-devel with wxWidgets 3.1.3.

It works on Linux, but crashes in glib-2.0, on something pango-related, after giving lots of "this is not a G_OBJECT" warnings.
Have you tried with wxGTK?

R.

**Comment by mojca** [comment]
The build issues definitely come from Cocoa, but no, I didn't yet try from GTK. Maybe I could check that next.

**Comment by RJVB** [comment]
I'm not seeing build issues, but something more annoying: a runtime issue that looks like it must come from something naughty done in Audacity's wx patches. I didn't get that (in the same Audacity code) with the previous wxWidgets 3.2 build. If I don't get to the bottom of it soon I'm reverting my wxWidgets upgrade...

**Comment by RJVB** [comment]
You've been looking into Audacity's custom wxWidgets implementation, right? What if the issue I'm seeing comes from mixing different implementations of certain class(es)?
The crash I'm seeing comes from glib2 accessing a private struct member (g_type_instance.g_class) under a wxWidgets member variable (wxWindowDCImpl::m_context, in the class ctor) that is a non-null value which is not a valid address but is always the same for a given build. I have a hunch that would fit with using a customised class definition that e.g. adds a member variable elsewhere but after all the stock ones (IF even that is acceptable...).

If that is indeed the reason then we're not safe from a similar issue on Mac too.

I'm now trying to figure out if I can put a breakpoint on the printing of error messages like
```
(Audacity:32030): GLib-GObject-CRITICAL **: 22:25:39.993: g_object_unref: assertion 'G_IS_OBJECT (object)' failed
(Audacity:32030): GLib-GObject-CRITICAL **: 22:25:39.998: g_object_ref: assertion 'G_IS_OBJECT (object)' failed
```

because those must be triggered by the same kind of issue that leads to the crash (annoyingly glib uses a macro around the assert so the breakpoint cannot be simply in g_object_ref or g_object_unref).

Can you share here if building against a stock wxWidgets was discussed during the workshop, and what came out of that discussion?

**Comment by mojca** [comment]
The agreement was to simply try to create a pull request to wxWidgets and get rid of the internal patches (they mostly didn't do that yet just for the lack of time).

Btw: I didn't test the builds against the patched wxWidgets, but I got a briefing about how to test the accessibility, so at least I now know what to look for. So anything I reported is from "stock wxWidgets".

**Comment by mojca** [comment]
Now, wxWidgets developers fixed the thing that was de-facto broken in 3.1.3. The rest of the problems come from Audacity using some probably-ment-to-be-private files and classes, and those have changes between 3.1.1 (the last one they tested with) and 3.1.3.

We can either postpone updating wxWidgets altogether, or we can make Audacity depend on 3.0 for now (if that still works) and switch to 3.1.3 only once Audacity releases the changes for compatibility with 3.1.3 (it's possible that they might do it within 10 days or so)? Or we can make a temporary wxWidgets-3.2-audacity port like we do for wxPython-3.0.

**Comment by ra1nb0w** [comment]
my point of view:
- commit this PR; we need to include this [patch](https://github.com/wxWidgets/wxWidgets/commit/294c8a6b23de686436a36f076de6a004a110a4d1) ?
- make Audacity depend on 3.0 (if it works)
This simplify the progress with Audacity.

**Comment by RJVB** [comment]
On Wednesday November 06 2019 04:27:09 rainbow wrote:

>- make Audacity depend on 3.0 (if it works)
>This simplify the progress with Audacity.

Or make a (temporary) port:wxWidgets-3.1 that is in fact a late checkout from not long before 3.1.3 (I know 3.1.2.916 works; it is what I have in MacStrop; works with Audacity 2.3.2.361).

**Comment by cjones051073** [comment]
Just out of curiosity, why is the port here called `wxWidgets-3.2` if it is in fact installing a 3.1.x release ??

**Comment by mojca** [comment]
> Just out of curiosity, why is the port here called `wxWidgets-3.2` if it is in fact installing a 3.1.x release ??

I started the numbering short before the 3.0 release when I had to decide between:
* adding a port wxWidgets-2.9, just to rename it, change all dependent ports and variants and obsolete it shortly after
* lying about the version a tiny bit as the versions 2.9.x were in fact just alpha releases of 3.0 (and 3.1.x are just alpha releases of 3.2)

**Comment by RJVB** [comment]
Vaguely and diagonally related: @Mojca, have you been following the development of wxQt? It seems continue steadily, most of the samples now build and run more or less correctly.

**Comment by ra1nb0w** [comment]
we should not lost the point.

added to PR the [patch](https://github.com/wxWidgets/wxWidgets/commit/294c8a6b23de686436a36f076de6a004a110a4d1) about `datatransfer.h`

**Comment by mojca** [comment]
I suggest to resolve this by the end of the weekend. My plan would be to create a new subport `wxWidgets-3.2-audacity` which would also include the non-upstreamed accessibility patches written by Audacity developers, and we can keep the version at "whichever we want" (I do expect further changes to both Audacity and wxWidgets which might need testing anyway).

In the meantime, can someone please patch `recode` (maintained by @ryandesign) to build correctly?

**Comment by RJVB** [comment]
> My plan would be to create a new subport `wxWidgets-3.2-audacity`

I'd suggest against that. Back when VLC couldn't build against FFmpeg 3+ (or 4?) I was allowed to introduce a port:ffmpeg-VLC *as a subport* but had to make certain that no one else could use it (too easily at least).

Frankly, it makes much more sense to me to introduce a more or less temporary wxWidgets-3.1 port from the git repo @ v3.1.2-916-gb383d539b5 (or whatever newer commit still works with Audacity). Even if in the end it's a what's in a name game...
FWIW, I'm working on a port:wxWidgets-3.1 as we speak, Linux first, then I'll make a Mac-only version.

I suppose the patches mentioned can have a wider interest than for Audacity alone; if so wxWidgets-3.1 can be the perfect test-bed for upstreaming them (but how non-trivial will it be to make Audacity build without any of its own wx extensions, as long as we already have a ton of patches of our own that could interfere?!)

**Comment by mojca** [comment]
We need to start upstreaming our patches, and Audacity needs to upstream theirs. This is why it would be useful to have some background about why the individual patches were needed. We cannot upstream the patch that replaces the python version, but the one to add British translations is something massive that we should explain and upstream, along with others.

Please note that the wxWidgets 3.1 upgrade is probably not really needed or urgent in any way, but it's nice to be up to date, and if Audacity has some special needs, I consider it to be an acceptable workaround. We already had wxPerl-3.0 in the past and we still have wxPython-3.0 just to be able to handle the incompatibilities. Also, if we decide to go a few commits up or down to be able to test Audacity, we won't need to worry about revbumping a gazillion other ports. We could also play a bit with CMake in Audacity-specific wxWidgets version, if needed. Audacity will try to migrate to CMake eventually, and that might be the perfect moment to test.

wxQt: No, I didn't test it yet, but to some extent it might combine the worst of both worlds: inability to build it on older macs (unless they provide very good compatibility with different Qt versions) and inability to use advanced Qt features that one might want. It might be a good alternative for when certain Aqua support is broken though, but it would also be useful to some extent to have gtk3 with quartz working properly.

**Comment by RJVB** [comment]
I just realised something: why does wxWidgets-3.2 install into a 3.1 folder? If 3.1 is just a beta for 3.2 then it should go into 3.2 IMHO - esp. if we do add a more-or-less temporary wxWidgets-3.1 port that will indeed be an actual 3.1 .

>This is why it would be useful to have some background about why the individual patches were needed.

I added some (rather minimal) descriptions to the patchfiles (where I couldn't do so in the Audacity portfile).

>Please note that the wxWidgets 3.1 upgrade is probably not really needed or urgent in any way

Audacity does prefer 3.1x nowadays, IIRC.

>wxQt: No, I didn't test it yet, but to some extent it might combine the worst of both worlds: inability to build it on older mac

I didn't check but it works with Qt 5.9.
Once the kinks are worked out in it and apps start recognising it I think it should provide a best of both worlds because it would allow wxWidgets development to concentrate on wx specifics rather than on platform specifics. Just as with the GTk-based versions you also get access to the various Qt tricks to use a different widget style (IMHO no existing cross-platform middleware does a good-looking job with the native Aqua style unless in apps designed specifically for Mac).

> gtk3 with quartz working properly.

I find that GTk3 is going more and more overboard with UI simplification with each newer release. I don't know if this is as much the case with the Quartz version as it is with the X11 version, but e.g. file dialogs under wxWidgets-3.2+gtk3 are frustratingly crippled (and wasting screen real estate). With Qt that problem doesn't exist.

Side-note: is there still actually anything called Quartz in the current Mac OS technologies?

File: graphics/wxWidgets-3.2/Portfile [hunk 1]
```diff
@@ -50,7 +50,8 @@ depends_lib-append  port:jpeg \
 depends_run         port:wxWidgets-common \
                     port:wxWidgets_select
 
-patchfiles          patch-configure.diff
+patchfiles          patch-configure.diff \
```
Comment (RJVB) [inline graphics/wxWidgets-3.2/Portfile]: Is `patch-configure` still required? From what I understand it's for the wxGTK build variant; surely the "native" build shouldn't find OpenGL from $prefix?

Comment (ra1nb0w) [reply graphics/wxWidgets-3.2/Portfile]: I lost this, sorry! seems to works fine without. removed.
thanks for the good catch.

**Comment by ra1nb0w** [comment]
rebased

**Comment by l2dy** [comment]
@mojca Is this good to merge after resolving merge conflicts?

**Comment by mojca** [comment]
Does Audacity (and Phoenix / wxPython 4.0, whatever it's called) still compile after these changes / do we have the working solution for it? If it's broken, let's make it a deadline to clean it up by the end of this week, and merge the PR either way.

@RJVB has had a bit more time to continue playing with Audacity than I did. I planned to finish this in particular when the Phoenix PR came up, but I somehow struggle with time a tiny bit.

**Comment by RJVB** [comment]
I haven't continued to try to get Audacity to build against the newer wxWidgets, partly because upstream seem to acknowledge that there are issues which they intend to address later.

For my own version of port:audacity I did what I suggested: it now depends on a port:wxWidgets-3.1 for the time being, which is actually v3.1.2.916 (b383d539b5867828049d2e98d305867442f06003). Not ideal, but it would allow port:wxWidgets-3.2 to move on.
With Leland I have been working on the CMake build system (only on Linux for me, for the time being). There's a strange regression, a crash that happens only when building through CMake; we're focussing on that now.

**Comment by ra1nb0w** [comment]
rebased. we should merge.

**Comment by michaelld** [comment]
> rebased. we should merge.

If there's any time to merge this, it is now with it being hot off the rebase! My vote is to do the merge. @mojca knows the most about wx itself and related ports ... any reason you see to not merge?

**Comment by ra1nb0w** [comment]
at this point we should merge (also to try to fix wxpython4) and then see what happen.

**Comment by emcrisostomo** [comment]
@mojca, any opinion about this?

**Comment by RJVB** [comment]
Mojca has been MIA for about a week now...

**Comment by l2dy** [comment]
Maintainer timeout, so merging it anyway. Thanks everyone!

**Pull request merged**